### PR TITLE
Add support for VectorNonlinearFunction

### PR DESCRIPTION
**Pull request by odow**
Tests won't pass because requires new tag of MOI.

But this is working with `PATHSolver#od/nlp-expr`

```julia
julia> using JuMP, PATHSolver

julia> model = Model(PATHSolver.Optimizer)
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: Path 5.0.03

julia> @variable(model, 0 <= x[1:4] <= 10)
4-element Vector{VariableRef}:
 x[1]
 x[2]
 x[3]
 x[4]

julia> @constraints(model, begin
                              -x[3]^2 -  x[4]   + 2 ⟂ x[1]
                               x[3]^3 - 2x[4]^2 + 2 ⟂ x[2]
           x[1]^5 -  x[2]   + 2x[3]   - 2x[4]   - 2 ⟂ x[3]
           x[1]   + 2x[2]^3 - 2x[3]   + 4x[4]   - 6 ⟂ x[4]
       end)
([-x[3]² - x[4] + 2, x[1]] ∈ MathOptInterface.Complements(2), [((x[3] ^ 3) - (2 x[4]²)) + 2.0, +(x[2])] ∈ MathOptInterface.Complements(2), [((((x[1] ^ 5) - x[2]) + (2 x[3])) - (2 x[4])) - 2.0, +(x[3])] ∈ MathOptInterface.Complements(2), [(((x[1] + (2.0 * (x[2] ^ 3))) - (2 x[3])) + (4 x[4])) - 6.0, +(x[4])] ∈ MathOptInterface.Complements(2))

julia> optimize!(model)
Path 5.0.03 (Fri Jun 26 09:58:07 2020)
Written by Todd Munson, Steven Dirkse, Youngdae Kim, and Michael Ferris


Crash Log
major  func  diff  size  residual    step       prox   (label)
    0     0             9.3949e+00             0.0e+00 (f[    4])
    1     4     3     2 7.7898e+00  4.1e-01    0.0e+00 (f[    1])
    2     5     1     3 1.4635e+00  1.0e+00    0.0e+00 (f[    3])
    3     6     0     3 1.1831e+00  1.0e+00    0.0e+00 (f[    3])
pn_search terminated: no basis change.

Major Iteration Log
major minor  func  grad  residual    step  type prox    inorm  (label)
    0     0     7     4 1.1831e+00           I 0.0e+00 1.2e+00 (f[    3])
    1     1     8     5 1.0633e+00  1.0e+00 SO 0.0e+00 1.1e+00 (f[    3])
    2     1     9     6 9.7034e-01  1.0e+00 SO 0.0e+00 9.7e-01 (f[    3])
    3     3    10     7 1.5017e+00  1.0e+00 SO 0.0e+00 1.2e+00 (f[    4])
    4     2    11     8 1.4820e+00  1.0e+00 SO 0.0e+00 1.2e+00 (f[    4])
    5     1    12     9 1.4756e+00  1.0e+00 SO 0.0e+00 1.2e+00 (f[    4])
    6     1    13    10 1.4483e+00  1.0e+00 SO 0.0e+00 1.2e+00 (f[    4])
    7     1    14    11 1.3317e+00  1.0e+00 SO 0.0e+00 1.1e+00 (f[    4])
    8     1    15    12 1.0594e+00  1.0e+00 SO 0.0e+00 1.0e+00 (f[    4])
    9     1    16    13 5.7505e-01  1.0e+00 SO 0.0e+00 5.7e-01 (f[    4])
   10     1    17    14 7.6079e-02  1.0e+00 SO 0.0e+00 7.6e-02 (f[    4])
   11     1    18    15 1.0305e-03  1.0e+00 SO 0.0e+00 1.0e-03 (f[    4])
   12     1    19    16 1.8656e-07  1.0e+00 SO 0.0e+00 1.9e-07 (f[    4])

Major Iterations. . . . 12
Minor Iterations. . . . 15
Restarts. . . . . . . . 0
Crash Iterations. . . . 3
Gradient Steps. . . . . 0
Function Evaluations. . 19
Gradient Evaluations. . 16
Basis Time. . . . . . . 0.000089
Total Time. . . . . . . 0.043275
Residual. . . . . . . . 1.865551e-07

julia> value.(x)
4-element Vector{Float64}:
 1.2847523472583888
 0.972916458993744
 0.909376156748673
 1.173035005537013
```

cc @mitchphillipson we're very close

### CODE REVIEW

**Comment by odow** [comment]
Here's a more complicated example
```Julia
julia> pi = 0.01
0.01

julia> L = 0.5
0.5

julia> γ = 0.02
0.02

julia> σ = 0.5
0.5

julia> ρ = 1 - 1 / 0.5
-1.0

julia> using JuMP, PATHSolver

julia> model = Model(PATHSolver.Optimizer)
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: Path 5.0.03

julia> @variable(model, EU, start = 1)
EU

julia> @variable(model, EV, start = 1)
EV

julia> @variable(model, C_G, start = 1)
C_G

julia> @variable(model, C_B, start = 1)
C_B

julia> @variable(model, K, start = 1)
K

julia> @constraints(model, begin
           EU - ((1 - pi) * C_G^ρ / ρ + pi * C_B^ρ / ρ)                       ⟂ EU
           EV - 100 * (((1 - pi) * C_G^ρ + pi * C_B^ρ)^(1 / ρ) - 1)           ⟂ EV
           C_G - (1 - γ * K)                                                  ⟂ C_G
           C_B - (1 - L + (1 - γ) * K)                                        ⟂ C_B
           γ * ((1 - pi) * C_G^(ρ - 1) + pi * C_B^(ρ - 1)) - pi * C_B^(ρ - 1) ⟂ K
       end)
([EU - (((0.99 * (C_G ^ -1.0)) / -1.0) + ((0.01 * (C_B ^ -1.0)) / -1.0)), +(EU)] ∈ MathOptInterface.Complements(2), [EV - (100.0 * ((((0.99 * (C_G ^ -1.0)) + (0.01 * (C_B ^ -1.0))) ^ -1.0) - 1.0)), +(EV)] ∈ MathOptInterface.Complements(2), [C_G + 0.02 K - 1, C_G] ∈ MathOptInterface.Complements(2), [C_B - 0.98 K - 0.5, C_B] ∈ MathOptInterface.Complements(2), [(0.02 * ((0.99 * (C_G ^ -2.0)) + (0.01 * (C_B ^ -2.0)))) - (0.01 * (C_B ^ -2.0)), +(K)] ∈ MathOptInterface.Complements(2))

julia> optimize!(model)
Path 5.0.03 (Fri Jun 26 09:58:07 2020)
Written by Todd Munson, Steven Dirkse, Youngdae Kim, and Michael Ferris


Crash Log
major  func  diff  size  residual    step       prox   (label)
    0     0             2.2871e+00             0.0e+00 (f[    1])
    1     1     0     5 5.2034e-01  1.0e+00    0.0e+00 (f[    2])
pn_search terminated: no basis change.

Major Iteration Log
major minor  func  grad  residual    step  type prox    inorm  (label)
    0     0     2     2 5.2034e-01           I 0.0e+00 5.2e-01 (f[    2])
    1     1     3     3 1.0494e-01  1.0e+00 SO 0.0e+00 1.0e-01 (f[    2])
    2     1     4     4 1.8864e-02  1.0e+00 SO 0.0e+00 1.9e-02 (f[    2])
    3     1     5     5 6.1192e-04  1.0e+00 SO 0.0e+00 6.1e-04 (f[    2])
    4     1     6     6 5.9024e-07  1.0e+00 SO 0.0e+00 5.9e-07 (f[    2])

Major Iterations. . . . 4
Minor Iterations. . . . 4
Restarts. . . . . . . . 0
Crash Iterations. . . . 1
Gradient Steps. . . . . 0
Function Evaluations. . 6
Gradient Evaluations. . 6
Basis Time. . . . . . . 0.000038
Total Time. . . . . . . 0.000949
Residual. . . . . . . . 5.902418e-07

julia> value(EU), value(EV), value(C_G), value(C_B), value(K)
(-1.0083430730335787, -0.8274042096088718, 0.9959051992930924, 0.7006452346384702, 0.20474003534537774)
```

File: src/nlp_expr.jl [hunk 1]
```diff
@@ -869,3 +869,35 @@ macro register(model, op, args...)
     )
     return Expr(:(=), esc(op), rhs)
 end
+
+
```
Comment (odow) [inline src/nlp_expr.jl:873]:
```suggestion
```

File: docs/src/tutorials/nonlinear/complementarity.jl [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+optimize!(model)
+Test.@test isapprox(value.(p), [0.225, 0.153, 0.126]; atol = 1e-3)  #src
+value.(p)
+
+# ## Expected utility of insurannce
+
+# This example is taken from a lecture of the course AAE706, given by Thomas F.
+# Rutherford at the University of Wisconsin, Madison. It models the expected
+# coverage of insurance `K` that a rational actor would obtain to insure a risk
+# that occurs with probability `pi` and results in a loss of `L`.
```
Comment (odow) [inline docs/src/tutorials/nonlinear/complementarity.jl:86]: @mitchphillipson when we merge nonlinear support I'll also add these examples to the documentation. Did I get this right? How should I cite the insurance example?

File: docs/Project.toml [hunk 3]
```diff
@@ -48,6 +49,7 @@ JSONSchema = "1"
 Literate = "2.8"
 MathOptInterface = "=1.18.0"
 MultiObjectiveAlgorithms = "=1.0.0"
+PATHSolver = "1"
```
Comment (odow) [inline docs/Project.toml:52]: TODO: update with released tag

File: docs/src/tutorials/nonlinear/complementarity.jl [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+optimize!(model)
+Test.@test value.([w, x, y, z]) ≈ [2.8, 0.0, 0.8, 1.2]  #src
+value.([w, x, y, z])
+
+# ## Transportation
+
+# This is example is a reformulation of the transportation problem from Chapter
+# 3.3 of Dantzig, G.B. (1963). _Linear Programming and Extensions_. Princeton
+# University Press, Princeton, New Jersey. It is based on the GAMS model
+# [gamslib_transmcp](https://www.gams.com/latest/gamslib_ml/libhtml/gamslib_transmcp.html).
```
Comment (odow) [inline docs/src/tutorials/nonlinear/complementarity.jl:54]:
```suggestion
# [`gamslib_transmcp`](https://www.gams.com/latest/gamslib_ml/libhtml/gamslib_transmcp.html).
```

File: docs/src/tutorials/nonlinear/complementarity.jl [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        [i in plants, j in markets], w[i] + cost[i=>j] - p[j] ⟂ x[i, j]
+        [i in plants], capacity[i] - sum(x[i, :]) ⟂ w[i]
+        [j in markets], sum(x[:, j]) - demand[j] ⟂ p[j]
+    end
+)
+optimize!(model)
+Test.@test isapprox(value(p["new-york"]), 0.225; atol = 1e-3)  #src
+value.(p)
+
+# ## Expected utility of insurannce
```
Comment (odow) [inline docs/src/tutorials/nonlinear/complementarity.jl:84]:
```suggestion
# ## Expected utility of insurance
```

**Comment by odow** [comment]
Preview https://jump.dev/JuMP.jl/previews/PR3450/tutorials/nonlinear/complementarity/

File: src/nlp_expr.jl [hunk 6]
```diff
@@ -869,3 +869,34 @@ macro register(model, op, args...)
...
+function moi_function_type(::Type{<:Vector{<:AbstractJuMPScalar}})
+    return MOI.VectorNonlinearFunction
+end
+
+function moi_function(f::Vector{<:AbstractJuMPScalar})
+    return MOI.VectorNonlinearFunction(map(moi_function, f))
+end
+
+function MOI.VectorNonlinearFunction(f::Vector{<:AbstractJuMPScalar})
```
Comment (blegat) [inline src/nlp_expr.jl:900]: If you call `MOI.VectorNonlinearFunction([x + 1])` you will get a`MOI.VectorAffineFunction` won't you ? I would probably interchange with `moi_function`

Comment (odow) [reply src/nlp_expr.jl:900]: Hahaha. Oh. True dat.

**Pull request merged**